In [18]:
from __future__ import division
from collections import defaultdict
import re
import math
from unidecode import unidecode
import json
import itertools
import pandas as pd

spchars = re.compile('\`|\~|\!|\@|\#|\$|\%|\^|\&|\*|\(|\)|\_|\+|\=|\\|\||\{|\[|\]|\}|\:|\;|\'|\"|\<|\,|\>|\?|\/|\.|\-')
text_file=open('F:/Documents/NYU_S3/ML/NyuPoly-CS6923-Fall16-master2/NyuPoly-CS6923-Fall16-master/Homework 3/stopwords.txt', 'r')
ListText = text_file.read().split('\n')
#print ListText
TFList=[]
IDFList=defaultdict(float)
IDF=[]

In [19]:
#Function that recmove spacial character,convert charcters into lower case and remove special characters 
def make_text_parsable(text):
    # convert to unicode
    text = unidecode(text) #.decode('utf-8', 'ignore'))
    # convert text to lowercase
    text = text.lower()
    # remove special characters
    text = spchars.sub(" ", text)
    return(text)

In [20]:
#function to count the number of tokens which we splits by white space without removing stopwords.We have used defaultdict(int) 
#over here which helps  us for counting words
def count_wordswithstopwords(text, wc=None):
    if wc == None:
        wc = defaultdict(int)
    tokens = text.split(" ")
    #print len(tokens)
    #tokens1=[item for item in tokens if item not in ListText]
    for t in tokens:
        wc[t] += 1
    #print wc
    return(wc)

In [21]:
#function to count the number of tokens which we splits by white space with removing stopwords.We have used defaultdict(int) 
#over here which helps  us for counting words

def count_words(text, wc=None):
    if wc == None:
        wc = defaultdict(int)
    tokens = text.split(" ")
    #print len(tokens)
    tokens1=[item for item in tokens if item not in ListText]
    for t in tokens1:
        wc[t] += 1
    #print wc
    return(wc)

In [22]:
#function which opens file,calls the count_words,count_wordswithstopwords,make_text_parsable function
def extract_info(filename):
    import json
    wc = defaultdict(int)
    wcwith_stop_words=defaultdict(int)
    df = defaultdict(set)
    count = 0
    bloblist=[]
    with open(filename) as fin:
        for line in fin:
            #print line
            count += 1
            current = json.loads(line)
            #print current
            text = make_text_parsable(current["abstract"] + " " + \
                current["description"] + " " + current["title"])
            #computetf(wc,len(text.split(" ")),count)
            #print text
            wcwith_stop_words=count_wordswithstopwords(text,wcwith_stop_words)
            wc = count_words(text, wc)
            #print wc
            #computetf(wc,len(text.split(" ")),count)
    #print count
    sorted_wcwithstopwords = sorted(wcwith_stop_words.items(), key=lambda x: x[1], reverse=True)
    print "(Words,Counts) with stopwords in it"
    print "\n"
    print sorted_wcwithstopwords
    sorted_wc = sorted(wc.items(), key=lambda x: x[1], reverse=True)
    #print count
    return sorted_wc

In [23]:
#function to count idf
def computeidf(tw,filename):
    for word in range(0,len(tw)):
        cnt=0
        with open(filename) as fin:
            for line in fin:
                current = json.loads(line)
                text = make_text_parsable(current["abstract"] + " " + \
                current["description"] + " " + current["title"])
                if tw[word][0] in text.split():
                    cnt=cnt+1
                else:
                    continue
            IDFList[tw[word][0]]=math.log10(79/cnt)
            print ('idf for word',tw[word][0],'is ',math.log10(79/cnt))

In [24]:
#function to count tf
def computetf1(filename):
   with open(filename) as fin:
       cnt=1
       for line in fin:
           df = defaultdict(int)
           current = json.loads(line)
           print "\n"
           print "\n"
           print ("TF for each word in document",cnt)
           print "\n"
           print "\n"
           text = make_text_parsable(current["abstract"] + " " + \
                current["description"] + " " + current["title"])
           tokens = text.split(" ")
           tokens1=[item for item in tokens if item not in ListText]
           for word in tokens1:
               df[word]+=1
           for k,v in df.items():
               df[k]=df[k]/len(tokens1)
           print(df.items())
           TFList.append(df.items())
           cnt+=1

Part 1)Using Python For NLP

We want to find an answer to the following question:

What is the trend in python and data in Seattle?

Section A: Term Frequency with unigrams

Q1.	Without modifying the code, which is the most common word?

Ans): As can be seen from below code & output for "Words with count with stopwords included in it", the most common word without modifying the code is '' with count 1919.



In [25]:
tw= extract_info("F:/Documents/NYU_S3/ML/NyuPoly-CS6923-Fall16-master2/NyuPoly-CS6923-Fall16-master/Homework 3/data_file.txt")

#print tw



(Words,Counts) with stopwords in it


[('', 1919), ('the', 608), ('and', 540), ('of', 452), ('to', 403), ('a', 328), ('in', 239), ('data', 231), ('for', 192), ('is', 153), ('with', 133), ('this', 132), ('python', 122), ('on', 119), ('will', 117), ('how', 96), ('i', 94), ('we', 91), ('as', 89), ('from', 81), ('that', 80), ('learning', 77), ('talk', 75), ('be', 75), ('using', 68), ('are', 64), ('an', 62), ('you', 61), ('analysis', 61), ('can', 60), ('it', 58), ('or', 54), ('ll', 52), ('at', 51), ('more', 51), ('use', 48), ('some', 46), ('s', 44), ('by', 43), ('pandas', 42), ('have', 42), ('these', 41), ('machine', 39), ('models', 37), ('your', 37), ('what', 36), ('tools', 36), ('time', 34), ('but', 34), ('analytics', 33), ('code', 33), ('our', 33), ('science', 31), ('their', 31), ('they', 29), ('discuss', 29), ('based', 29), ('spark', 28), ('model', 28), ('new', 27), ('all', 27), ('also', 27), ('like', 26), ('well', 26), ('about', 26), ('into', 25), ('used', 24), ('tutorial', 23), ('lear

Q2.Notice that a lot of the more common words don’t give much info since they are what we can stopwords. Stop words are natural language words which have very little meaning, such as "and", "the", "a", "an", and similar words. Now, look at the file called stopwords and modify tools.py so that it excludes stopwords in the counting. What are the 3 most common words now?

Ans): As can be seen from below code & output for "Words with count with stopwords included in it",the 3 most common words after removing stop words are
('data', 231)
('with', 133)
('python', 122) 
accordingly in descending order of count




In [26]:
print "\n"
print "(Words,Counts) without stopwords in it"
print "\n"
print tw

#note I have run the script before adding the 3 more words that we have asked in the question 3. So before adding that I'm getting the below output for question 2



(Words,Counts) without stopwords in it


[('data', 231), ('with', 133), ('python', 122), ('learning', 77), ('talk', 75), ('analysis', 61), ('ll', 52), ('more', 51), ('use', 48), ('pandas', 42), ('machine', 39), ('models', 37), ('tools', 36), ('time', 34), ('analytics', 33), ('code', 33), ('science', 31), ('discuss', 29), ('based', 29), ('spark', 28), ('model', 28), ('new', 27), ('tutorial', 23), ('learn', 23), ('both', 23), ('work', 21), ('techniques', 21), ('not', 21), ('web', 21), ('methods', 20), ('library', 20), ('performance', 19), ('world', 19), ('coming', 19), ('algorithm', 19), ('software', 19), ('algorithms', 18), ('numpy', 18), ('building', 18), ('soon', 18), ('series', 17), ('applications', 17), ('statistical', 17), ('space', 17), ('different', 17), ('deep', 17), ('build', 17), ('libraries', 17), ('examples', 17), ('intel', 16), ('describe', 16), ('large', 16), ('processing', 16), ('users', 16), ('project', 15), ('implementation', 15), ('people', 15), ('language', 15), ('o

Q3. Given that this is a list of talks about data and python, do you think that there are some words that don’t add much value? Name 3 of them and add them to the stopwords list.

Ans): I have added three words in my stopword list 1)data 2)python 3)more
As the documents are related to Machine learning and the talks in list is about data and python,The most common words relating to this subject are Data and Python.We can include these words in our stopword list as they are the most common terms in this area and they are not adding much value in terms of our experiment
I've added "more" in the stopword list as it is most common stop word we found while reading or analyzing document

**Reference
http://www.ranks.nl/stopwords

Q4. Now run your code again with the new list of stopwords. What’s the most common word now?

A: As can be seen from below code & output, the most common word are 'with,'learning' and 'talk' with counts 133, 77 and 75 respectively.

In [10]:
print tw

#note after adding python,data,more in stop word list I'm getting the below output

[('with', 133), ('learning', 77), ('talk', 75), ('analysis', 61), ('ll', 52), ('use', 48), ('pandas', 42), ('machine', 39), ('models', 37), ('tools', 36), ('time', 34), ('analytics', 33), ('code', 33), ('science', 31), ('discuss', 29), ('based', 29), ('spark', 28), ('model', 28), ('new', 27), ('tutorial', 23), ('learn', 23), ('both', 23), ('work', 21), ('techniques', 21), ('not', 21), ('web', 21), ('methods', 20), ('library', 20), ('performance', 19), ('world', 19), ('coming', 19), ('algorithm', 19), ('software', 19), ('algorithms', 18), ('numpy', 18), ('building', 18), ('soon', 18), ('series', 17), ('applications', 17), ('statistical', 17), ('space', 17), ('different', 17), ('deep', 17), ('build', 17), ('libraries', 17), ('examples', 17), ('intel', 16), ('describe', 16), ('large', 16), ('processing', 16), ('users', 16), ('project', 15), ('implementation', 15), ('people', 15), ('language', 15), ('open', 15), ('most', 15), ('available', 15), ('uses', 15), ('source', 15), ('application',

Section B: Using TF-IDF

Q5. Compute TF-IDF for the corpus. Note that you will need to compute a separate score for TF per document. Again in our file we have 79 documents.

A: Following functions will compute TF, IDF & TF-IDF for the corpus. 
    1st function will give results of IDF values for words.
    2nd function will give results of TF values for each documents in corpus.
    3rd function will give results of IDF*TF value.

In [11]:
print "\n"
print "(IDF for each Word)"
print "\n"
computeidf(tw,"F:/Documents/NYU_S3/ML/NyuPoly-CS6923-Fall16-master2/NyuPoly-CS6923-Fall16-master/Homework 3/data_file.txt")
#print IDFList.items()
print "\n"




(IDF for each Word)


('idf for word', 'with', 'is ', 0.21638585391485418)
('idf for word', 'learning', 'is ', 0.5754077965565222)
('idf for word', 'talk', 'is ', 0.29556709996247904)
('idf for word', 'analysis', 'is ', 0.5552044104682352)
('idf for word', 'll', 'is ', 0.5552044104682352)
('idf for word', 'use', 'is ', 0.48265374331962346)
('idf for word', 'pandas', 'is ', 0.8976270912904415)
('idf for word', 'machine', 'is ', 0.7514990556122034)
('idf for word', 'models', 'is ', 0.8184458452428166)
('idf for word', 'tools', 'is ', 0.6935071086345167)
('idf for word', 'time', 'is ', 0.6423545861871354)
('idf for word', 'analytics', 'is ', 0.8976270912904415)
('idf for word', 'code', 'is ', 0.6671781699121675)
('idf for word', 'science', 'is ', 0.8562344061322164)
('idf for word', 'discuss', 'is ', 0.6423545861871354)
('idf for word', 'based', 'is ', 0.6671781699121675)
('idf for word', 'spark', 'is ', 1.295567099962479)
('idf for word', 'model', 'is ', 0.7836837389836047)
('idf for w

In [12]:
print "(TF for each Document)"
print "\n"
computetf1("F:/Documents/NYU_S3/ML/NyuPoly-CS6923-Fall16-master2/NyuPoly-CS6923-Fall16-master/Homework 3/data_file.txt")
merged = list(itertools.chain.from_iterable(TFList))

(TF for each Document)






('TF for each word in document', 1)




[('breakfast', 0.5), ('registration', 0.5)]




('TF for each word in document', 2)




[('code', 0.0375), ('ve', 0.0125), ('just', 0.0125), ('century', 0.0125), ('helped', 0.025), ('bring', 0.025), ('shape', 0.0125), ('computer', 0.05), ('platform', 0.0125), ('learning', 0.0125), ('100', 0.0125), ('bold', 0.0125), ('foundational', 0.0125), ('districts', 0.0125), ('21st', 0.0125), ('field', 0.025), ('with', 0.0375), ('support', 0.0125), ('better', 0.0125), ('majority', 0.0125), ('online', 0.0125), ('courses', 0.0125), ('8', 0.0125), ('public', 0.0125), ('wants', 0.0125), ('keynote', 0.0125), ('description', 0.0125), ('10', 0.0125), ('students', 0.025), ('enrolled', 0.0125), ('free', 0.0125), ('released', 0.0125), ('revolutionizing', 0.0125), ('student', 0.0125), ('70', 0.0125), ('schools', 0.0375), ('org', 0.025), ('teach', 0.0125), ('america', 0.0125), ('opportunity', 0.025), ('change', 0.0125), ('partnered', 0.0125)

In [13]:
dataframe= pd.DataFrame(merged,columns=["Word","TF"])
dataframe=pd.DataFrame(dataframe.groupby(['Word'])['TF'].max(),columns=["TF"])
dataframe=dataframe.reset_index()
IDF.append(IDFList.items())
#print IDF

In [14]:
idfdataframe=pd.DataFrame(list(itertools.chain.from_iterable(IDF)),columns=["Word","IDF"])
idfdataframe=idfdataframe.sort(['Word'],ascending=True)
idfdataframe=idfdataframe.reset_index(drop=True)
#print(idfdataframe)


c:\python27\lib\site-packages\ipykernel\__main__.py:2: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  from ipykernel import kernelapp as app


In [15]:
#print(dataframe)

In [16]:
result=idfdataframe.merge(dataframe, on='Word', how='inner')
#print result

Q6. Rank each unique word by their TF-IDF score. Which one has the highest score? Use only the highest for each word. So if a word appears twice only use the highest score.

A: Printed the data frame which is sorted in descending order of TF-IDF values .The words which has highest score for TF-IDF values are
1)lunch with value of TF-IDF 1.420506 
2)breakfast with value of TF-IDF 1.420506

In [17]:
result['tf*idf']=result['IDF']*result['TF']
result=result.sort(['tf*idf'],ascending=False)
print result

                  Word       IDF        TF    tf*idf
1362             lunch  1.420506  1.000000  1.420506
293          breakfast  1.420506  1.000000  1.420506
1319         lightning  1.897627  0.500000  0.948814
1892      registration  1.596597  0.500000  0.798299
2257             talks  1.596597  0.500000  0.798299
292              break  1.420506  0.500000  0.710253
2111            snacks  1.420506  0.500000  0.710253
2122              soon  0.943385  0.400000  0.377354
415             coming  0.897627  0.400000  0.359051
2129             space  1.596597  0.170213  0.271761
1261           keynote  1.198657  0.200000  0.239731
1436            mining  1.897627  0.125000  0.237203
548               dark  1.897627  0.125000  0.237203
1415             memex  1.897627  0.125000  0.237203
1187             intel  1.897627  0.122137  0.231771
205              azure  1.897627  0.111111  0.210847
1392        matplotlib  1.420506  0.142857  0.202929
2166             state  1.420506  0.142857  0.

c:\python27\lib\site-packages\ipykernel\__main__.py:2: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  from ipykernel import kernelapp as app


Q7. Is the space '' still there? As one of the words. If so how can you remove it?

Ans-7)No the space '' is not there after removing stop words from list.It was first there at the time of answering question 1). To remove '' ,I have added one blank line in stopwords file and split the stopwords file by new line character which adds '' in the list of stop words

Q8. Paste the output for both Section A and Section B.

Ans-8)Printed all the output for Section A and Section B




***Part 2-Pen and Pencil


Q 2: Consider an information retrieval task, where the problem is to predict whether a given reader will be interested in book x. The book should be labeled + if the reader will be interested in it, and otherwise. The following table shows a small training set for this problem, consisting of 7 books. It also shows the predictions made by a hypothesis h on these books. The column labeled “r” gives the label of the book in the training set, and the column labeled y gives the prediction of h.
Calculate the recall and precision of h. (If we imagine using h to decide which books to give to the reader, then Recall is calculated by considering the set of all books that are interesting to the reader, and determining what percentage of those books are given to the reader. Precision is calculated by considering all books that are given to the reader, and determining what percentage of those books are interesting to the reader.)

![title](img/1.jpg)


Q-3) Let h be a hypothesis for a binary classification problem which assigns a probability or score to each example x in a test set X. We can obtain different points on the ROC curve for h by changing the threshold θ in the rule “if h(x) > θ, then classify x as +”. The number of points on the ROC curve is |X| + 1. In fact, the points on the ROC curve do not depend on the exact values assigned by h. To obtain the ROC curve, it is enough to give the ranking of the elements xt of the test set induced by the values of h on those elements. In other words, it is enough to list the elements xt in the test set in descending order of the value of h(xt). We hope that h will assign higher values to the positive examples than to the negative examples. Let X be a test set containing 6 examples. Three of the examples are labeled +, and three are labeled -. Suppose we list the examples xt in X in descending order of the value of h(x). (a) Suppose h ranks all 3 positive examples in X higher than all 3 negative examples in X (i.e., the highest values of h are on the positive examples). Graph the resulting ROC curve. What is the AUC (the area under the curve in the interval [0,1])? (b) Suppose h ranks all 3 negative examples in X higher than all 3 positive examples in X. Graph the resulting ROC curve. What is the AUC? (c) Suppose that the ranking assigned by h corresponds to the following list of labels, if we list the examples in descending order of rank: (+,+,−,+,−,−). So, the two highest values of h are given to two of the positive examples, the next highest to a negative example, etc. Graph the resulting ROC curve. What is the AUC?

![title](img/2.jpg)
![title](img/3.jpg)
![title](img/4.jpg)